### PR TITLE
MSC3925: m.replace aggregation with full event

### PR DESCRIPTION
**Pull request by benkuly**
[rendered](https://github.com/benkuly/matrix-spec-proposals/blob/main/proposals/3925-replace-aggregation-with-full-event.md)

Instead of always complaining, here is my first MSC :wink: 

[FCP tickyboxes](https://github.com/matrix-org/matrix-spec-proposals/pull/3925#issuecomment-1380150062)
Implementations: [synapse](https://github.com/matrix-org/synapse/pull/14811), [ele-web](https://github.com/matrix-org/matrix-js-sdk/pull/3045)

### CODE REVIEW

**Comment by turt2live** [comment]
@benkuly don't forget to sign off on the changes for if this eventually gets accepted please :)

**Comment by benkuly** [comment]
> @benkuly don't forget to sign off on the changes for if this eventually gets accepted please :)

Thank you, is done now.

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# MSC3925: m.replace aggregation with full event
+
+The Matrix DAG has immutable PDUs forming an auth chain. The client representation of this PDUs are simply called events.
+While this events also have been immutable until v1.3, since v1.4 they aren't.
```
Comment (jplatte) [inline proposals/3925-replace-aggregation-with-full-event.md:4]:
```suggestion
While these events also have been immutable until v1.3, since v1.4 they aren't.
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+
+
+## Security considerations
+
+
+
+## Unstable prefix
+
+I'm not sure, if we need an unsable prefix, because the aggregation would just be extended by additional fields.
```
Comment (jplatte) [inline proposals/3925-replace-aggregation-with-full-event.md:38]:
```suggestion
I'm not sure, if we need an unstable prefix, because the aggregation would just be extended by additional fields.
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+but only `event_id`, `origin_server_ts` and `sender` are included.
+Theoretically this is enough to get the replacing content, but when the event with the `event_id` cannot be found locally it needs to be fetched from the server.
+To prevent this additional call to the server, the `m.replace` aggregation should just contain the complete replacing event.
+
+The additional server call is already needed for encrypted events and would be saved by this proposal too.
+
+## Potential issues
+
+* There could be clients, which rely on the current behavior.
+* It is not as easy for clients like as in the current spec to get the current content of an event by just looking into `content.body`. While this true, it is also a relatively inconsistent behavior. Future replacements of the event would be rendered as "* new content". So the event with the replaced event does look different (without "*") despite the fact, that it is also replaced.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:26]: please wrap your text at 80 characters.

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# MSC3925: m.replace aggregation with full event
+
+The Matrix DAG has immutable PDUs forming an auth chain. The client representation of this PDUs are simply called events.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:3]: they are called events in the DAG as well. "PDU" isn't used much, but when it is, it's specific to transmitting individual events over the federation API.

The refererence to "auth chains" is irrelevant here too.

Generally I don't think this line is adding much, and I suggest you get rid of it.

```suggestion
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+While these events also have been immutable until v1.3, since v1.4 they aren't.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:4]: I think this is overstating the case. For one thing, it's always been possible to mutate events via redaction. Second, they are still immutable: it's just the server's representation of them which is mutating. (That representation is defined at https://spec.matrix.org/v1.4/client-server-api/#room-event-format, and isn't quite the same as the "raw event": note the presence of `event_id` for example)

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+When a client sends a `m.replace` relation, [the server should replace the content of the original event](https://spec.matrix.org/v1.4/client-server-api/#server-side-replacement-of-content).
+
+There are some issues with this requirement:
+* Changing the fundamental concept of immutable events is confusing. The server can respond with different event contents for the same `event_id`.
+* If an event with `m.replace` relation is deleted, the client would need to detect, if the original content was replaced and possibly needs to fetch the original content.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:9]: events can't be deleted. Do you mean `redacted`?

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+* There is an additional server call needed, when the replacing event is encrypted, because the server cannot replace the original event content.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:10]: I don't see how you can claim that an "additional server call" is needed, when it's unclear how a server can implement the spec as it is currently written anyway.

```suggestion
* Server cannot replace the original event content for encrypted events (because the 
  replacement content is inside the encrypted body).
  See [matrix-spec#1299](https://github.com/matrix-org/matrix-spec/issues/1299).
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+* There are also some other issues with this spec paragraph, which are discussed [here](https://github.com/matrix-org/matrix-spec/issues/1299)
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:11]: Spell them out for us, please? I'm not sure what other issues you are referring to.

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+In fact it is [already used](https://spec.matrix.org/v1.4/client-server-api/#server-side-aggregation-of-mreplace-relationships), 
+but only `event_id`, `origin_server_ts` and `sender` are included.
+Theoretically this is enough to get the replacing content, but when the event with the `event_id` cannot be found locally it needs to be fetched from the server.
+To prevent this additional call to the server, the `m.replace` aggregation should just contain the complete replacing event.
+
+The additional server call is already needed for encrypted events and would be saved by this proposal too.
+
+## Potential issues
+
+* There could be clients, which rely on the current behavior.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:25]:
```suggestion
* There could be clients which rely on the current behavior.
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+## Proposal
+
+Instead of replacing the original content of an event, servers should use the aggregation feature for it.
+In fact it is [already used](https://spec.matrix.org/v1.4/client-server-api/#server-side-aggregation-of-mreplace-relationships), 
+but only `event_id`, `origin_server_ts` and `sender` are included.
+Theoretically this is enough to get the replacing content, but when the event with the `event_id` cannot be found locally it needs to be fetched from the server.
+To prevent this additional call to the server, the `m.replace` aggregation should just contain the complete replacing event.
+
+The additional server call is already needed for encrypted events and would be saved by this proposal too.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:21]: the additional server call is needed because Synapse isn't implementing the spec, and it's not specific to encrypted events: https://github.com/matrix-org/synapse/issues/12793.

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+## Potential issues
+
+* There could be clients, which rely on the current behavior.
+* It is not as easy for clients like as in the current spec to get the current content of an event by just looking into `content.body`. While this true, it is also a relatively inconsistent behavior. Future replacements of the event would be rendered as "* new content". So the event with the replaced event does look different (without "*") despite the fact, that it is also replaced.
+
+## Alternatives
+
+
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:30]: we should probably mention @kegsay's alternative suggestion at https://github.com/matrix-org/matrix-spec/issues/1299#issuecomment-1290332433.

**Review by richvdh — CHANGES_REQUESTED** [review]
Thanks! I agree with the proposal, but I'm not sure I agree with all the reasons, and the proposal could do with some clarifications.

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+## Security considerations
+
+
+
+## Unstable prefix
+
+I'm not sure, if we need an unstable prefix, because the aggregation would just be extended by additional fields.
```
[the remark this answers is not in the document]
Comment (richvdh) [reply proposals/3925-replace-aggregation-with-full-event.md:37]: the fact that we're just adding extra fields is no reason that we would not need an unstable prefix.

Still, it's hard to see how an unstable prefix could be used here.

File: proposals/3925-replace-aggregation-with-full-event.md (hunk 6 above)
Comment (benkuly) [reply proposals/3925-replace-aggregation-with-full-event.md:8]: yep

File: proposals/3925-replace-aggregation-with-full-event.md (hunk 5 above)
Comment (benkuly) [reply proposals/3925-replace-aggregation-with-full-event.md:3]: While I still think immutablilty is relevant, I understand what you mean. In fact, in Trixnity, I circumvented all the cases, where some fields are removed by just adding them to the json before the deserialization. I've always disliked this, because it breaks good typing in most languages.
I removed the sentence.

File: proposals/3925-replace-aggregation-with-full-event.md (hunk 8 above)
Comment (benkuly) [reply proposals/3925-replace-aggregation-with-full-event.md]: There are actually no more, thank you.

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+When a client sends a `m.replace`
+relation, [the server should replace the content of the original event](https://spec.matrix.org/v1.4/client-server-api/#server-side-replacement-of-content).
+
+There are some issues with this requirement:
+
+* Changing the fundamental concept of mostly immutable events is confusing. The server
+  can respond with different event contents for the same `event_id`.
+* If an event with `m.replace` relation is redacted, the client would need to
+  detect, if the original content was replaced and possibly needs to fetch the
+  original content.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:12]:
```suggestion
* If an event with `m.replace` relation is redacted, clients need to
  detect if the original content was replaced, and possibly need to fetch the
  original content.
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 14]
```diff
@@ -0,0 +1,132 @@
+# MSC3925: m.replace aggregation with full event
+
+When a client sends a `m.replace`
+relation, [the server should replace the content of the original event](https://spec.matrix.org/v1.4/client-server-api/#server-side-replacement-of-content).
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:4]:
```suggestion
As [currently specified](https://spec.matrix.org/v1.4/client-server-api/#server-side-replacement-of-content),
servers should replace the content of events that have been replaced via an `m.replace`
relation.
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+* Changing the fundamental concept of mostly immutable events is confusing. The server
+  can respond with different event contents for the same `event_id`.
+* If an event with `m.replace` relation is redacted, the client would need to
+  detect, if the original content was replaced and possibly needs to fetch the
+  original content.
+* Server cannot replace the original event content for encrypted events (because the
+  replacement content is inside the encrypted body).
+  See [matrix-spec#1299](https://github.com/matrix-org/matrix-spec/issues/1299).
+
+There is a discussion about this spec paragraph [here](https://github.com/matrix-org/matrix-spec/issues/1299).
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:17]: this is mentioned just above
```suggestion
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+  // irrelevant fields not shown
+}
+
+```
+
+## Potential issues
+
+* There could be clients which rely on the current behavior.
+* It is not as easy for clients like as in the current spec to get the current
+  content of an event by just looking into `content.body`. While this is true, it
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:112]:
```suggestion
* It will be harder for clients to get the current content of an event: currently
  they can just looking at `content.body`. While this is true, it
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+  is also a relatively inconsistent behavior. Future replacements of the event
+  would be rendered as "* new content". So the event with the replaced event
+  does look different (without "*") despite the fact, that it is also replaced.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:115]: I'm afraid I'm not following this. When you say "Future replacements", do you mean events that arrive (over `/sync`) *after* the client receives the original event?

The term "rendered" refers the way a client takes the json and turns it into user-visible text. Is that really what you mean, or are you talking about the way a server returns the event?

> So the event with the replaced event does look different (without "*") 

I would certainly expect events *without* a replacement not to have a `*`, so this makes it sound as though events with replacements will look the same as those without? I don't think that's what you mean though?

Basically: please could you rewrite these two sentences.

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+implemented) then they see the right thing, and when someone does care about
+them, they
+can just inspect unsigned to present the "edits" dialog.
+
+## Security considerations
+
+## Unstable prefix
+
+I'm not sure if we need an unstable prefix, because the aggregation would just
+be extended by additional fields.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:132]:
```suggestion
No unstable prefix is currently proposed.
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+We could flip the event stack on it's head and have the outer event
+be the edit and then in unsigned have the base event being edited. Currently, it
+is the inverse where we have the original event and then in unsigned the newer
+event sits. That way, if someone doesn't care about edits (because not
+implemented) then they see the right thing, and when someone does care about
+them, they
+can just inspect unsigned to present the "edits" dialog.
+
+## Security considerations
+
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:128]:
```suggestion

None currently foreseen.

```

**Review by richvdh — CHANGES_REQUESTED** [review]

File: proposals/3925-replace-aggregation-with-full-event.md (hunk 17 above)
Comment (benkuly) [reply proposals/3925-replace-aggregation-with-full-event.md]: see https://github.com/matrix-org/matrix-spec-proposals/pull/3925/commits/5bfffb4ddf7c4da863ca71bdd1eaab26dd272239

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+## Alternatives
+
+We could flip the event stack on it's head and have the outer event
+be the edit and then in unsigned have the base event being edited. Currently, it
+is the inverse where we have the original event and then in unsigned the newer
+event sits. That way, if someone doesn't care about edits (because not
+implemented) then they see the right thing, and when someone does care about
+them, they
+can just inspect unsigned to present the "edits" dialog.
```
[the remark this answers is not in the document]
Comment (benkuly) [reply proposals/3925-replace-aggregation-with-full-event.md]: You are right. Thank you :)

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 21]
```diff
@@ -108,25 +108,27 @@ This is how the original event would look like after the replacement:
 ## Potential issues
 
 * There could be clients which rely on the current behavior.
-* It is not as easy for clients like as in the current spec to get the current
-  content of an event by just looking into `content.body`. While this is true, it
-  is also a relatively inconsistent behavior. Future replacements of the event
-  would be rendered as "* new content". So the event with the replaced event
-  does look different (without "*") despite the fact, that it is also replaced.
+* It will be harder for clients, which do not support replacing events to get
+  the current content of an event: currently they can just look at `content.body`.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:112]: Ok, point of English grammar: in English we don't use a comma here. If you include a comma, then "which" refers to the *whole* of the previous phrase ("It will be harder for clients") instead of just "clients". 

So, this is grammatically correct, though not what you mean here:

> "[Some action] will be harder for clients, which will [have some consequence]", eg "Displaying events will be harder for clients, which will be annoying for client developers"

In general, I think we use fewer commas in English than you think we do :)

```suggestion
* It will be harder for clients which do not support replacing events to get
  the current content of an event: currently they can just look at `content.body`.
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 22]
```diff
@@ -108,25 +108,27 @@ This is how the original event would look like after the replacement:
...
+  However, this leads to a relatively inconsistent behavior:
+
+  Assuming that we have an event A with the body `1`, which is first replaced by
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:115]:
```suggestion
  Assume that we have an event A with the body `1`, which is first replaced by
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 23]
```diff
@@ -108,25 +108,27 @@ This is how the original event would look like after the replacement:
...
+  the event A' with the new body `2` and then by the event A'' with the new body
+  `3`. The sync has been done right after A' und A''. Then the timeline
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:117]:
```suggestion
  `3`. A `/sync` is done right after A' and A''. Then the timeline
```

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 24]
```diff
@@ -108,25 +108,27 @@ This is how the original event would look like after the replacement:
...
+  would look like this: `2` -> `*3` instead of `1` -> `*2` -> `*3`. Although `2`
+  is a replaced body, it does not have a `*`, which looks to the user as it has 
+  not been replaced.
```
Comment (richvdh) [inline proposals/3925-replace-aggregation-with-full-event.md:120]: Sorry, still not following. Please can you break it down a bit more.  What do you mean by "the timeline would look like this: `2` -> `*3`" ? What does `->` represent here?

After A', the client receives a /sync like this:

```yaml
timeline:
  - { event_id: "A", content: {body: "1"} }
  - { event_id: "A'", content: {body: "* 2"} }
```

Right?

----

If I spent a long time staring at your proposal and thinking about what you are trying to say, I could probably figure it out.  But for a proposal to be accepted, everybody else needs to understand it, and you can't expect everyone to spend hours trying to reverse-engineer your proposal.

In other words, please remember your audience, and explain yourself clearly.

**Review by richvdh — CHANGES_REQUESTED** [review]

File: proposals/3925-replace-aggregation-with-full-event.md (hunk 24 above)
Comment (benkuly) [reply proposals/3925-replace-aggregation-with-full-event.md]: Sorry for the effort and all the commas.

Is this better? I think I would put that into the introduction, because it is also an issue of the current spec.

* Replacing the content can lead to inconsistent behavior on clients which don't
  support replacing events.
  Assume that we have this timeline on the server:
    - `{"event_id": "E1", "content": {"body": "1"}}`
    - `{"event_id": "E2", "content": {"body": "* 2", "m.new_content": {"body": "2"}}` // replaces `E1`

  A `/sync` is done after `E2`. We will have this timeline on a client:
    - `{"event_id": "E1", "content": {"body": "2"}}`
    - `{"event_id": "E2", "content": {"body": "* 2"}`

  Although `2` is a replaced body it does not have a `*`. This looks to the
  user as if it is the original content of the event.

Comment (richvdh) [reply proposals/3925-replace-aggregation-with-full-event.md]: Yes, that's much clearer, thanks. Good idea to put it in the introduction.

**Review by richvdh — APPROVED** [review]
looks good!

The next step is to get it implemented so that we can see it working.

**Comment by richvdh** [comment]
This now has server-side (https://github.com/matrix-org/synapse/pull/14811) and client-side (https://github.com/matrix-org/matrix-js-sdk/pull/3045) implementations. I believe it is ready for FCP.

@mscbot fcp merge

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+
+
+## Proposal
+
+The following two changes are proposed:
+1. The [server-side aggregation of `m.replace` relationships](https://spec.matrix.org/v1.4/client-server-api/#server-side-aggregation-of-mreplace-relationships)
+   is extended to be the entire content of the most recent replacement event, formatted
+   as described in [Room Event Format](https://spec.matrix.org/v1.4/client-server-api/#room-event-format).
+   This ensures that the client will always have the most recent edit without having to
+   fetch it from the server.
```
Comment (turt2live) [inline proposals/3925-replace-aggregation-with-full-event.md:38]: Sorry, I don't understand what is being described here: are we saying the replacement event applies to itself? (in which case, why bother with specifying new and old content separately?)

**Comment by turt2live** [comment]
@mscbot concern Unclear what is being proposed

File: proposals/3925-replace-aggregation-with-full-event.md (hunk 25 above)
Comment (richvdh) [reply proposals/3925-replace-aggregation-with-full-event.md:38]: @turt2live the current wording seems pretty clear to me (though I had a hand in writing it, so maybe I'm biased).

The current spec, as linked, says:

> The aggregation format of `m.replace` relationships gives the `event_id`, `origin_server_ts`, and` sender` of the most recent replacement event.

We're changing it so that it is the *whole* of that replacement event, rather than just three fields.

**Comment by turt2live** [comment]
@mscbot resolve Unclear what is being proposed

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+{
+  "event_id": "$edit_event",
+  "type": "m.room.message",
+  "content": {
+    "body": "* I really like *chocolate* cake",
+    "msgtype": "m.text",
+    "m.new_content": {
+      "body": "I really like *chocolate* cake",
+      "msgtype": "m.text",
+      "com.example.extension_property": "chocolate"
```
Comment (dbkr) [inline proposals/3925-replace-aggregation-with-full-event.md:70]: Is this here to demonstrate that newly added properties appear on the related event and not on the original? It would b e good to clarify as it's not super obvious why it's here.

File: proposals/3925-replace-aggregation-with-full-event.md [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+    }
+  }
+  // irrelevant fields not shown
+}
+
+```
+
+## Potential issues
+
+* There could be clients which rely on the current behavior.
```
Comment (dbkr) [inline proposals/3925-replace-aggregation-with-full-event.md:122]: Do we know of any in particular? What will the failure mode be in such clients and is it tolerable?

File: proposals/3925-replace-aggregation-with-full-event.md (hunk 26 above)
Comment (richvdh) [reply proposals/3925-replace-aggregation-with-full-event.md:70]: > Is this here to demonstrate that newly added properties appear on the related event and not on the original?

I think it's just an example, yes. Maybe it would be clearer just to remove it?

File: proposals/3925-replace-aggregation-with-full-event.md (hunk 27 above)
Comment (richvdh) [reply proposals/3925-replace-aggregation-with-full-event.md:122]: Well, element-web relied on it until https://github.com/matrix-org/matrix-js-sdk/pull/3045, and I suspect most other clients still do.

The failure mode is: suppose that some time ago, you have an event E, which was subsequently replaced by an event E'. Now, if you jump back to a bit of timeline that contains E but *not* E', then clients unaware of this change will show the original event E rather than the edited content E'.

My feelings are: (1) that is an edge-case; (2) it already happens for events other than `m.room.message` (including encrypted events) due to https://github.com/matrix-org/synapse/issues/12503.

All of this information should probably be in the MSC itself: @benkuly do you fancy updating it?

**Comment by richvdh** [comment]
Spec PR: https://github.com/matrix-org/matrix-spec/pull/1440

**Comment by richvdh** [comment]
merged!
